### ISSUE TITLE
Define index directory, effort or number of threads etc. through .clangd yaml file

### ISSUE THREAD

**Issue by soumen-ghosh** [clangd/clangd#2262]
I am setting up clang* tools for the 1st time. I am using 20.0.0git versions for the clang* tools (clangd, clang-format and clang-tidy) I have a question.

I see indexes are getting generated under `${HOME}/.cache/clangd/index directory`.

Since I have multiple sandboxes (perforce), I want indexes to be generated in the sandbox (project directory). What is the way to define that `.clangd` file? And is there a way to specify priority or effort or number of threads to use for indexing in the `.clangd` file?

I have defined `.clangd` file as following in the sandbox directory:

```
Hover:
  ShowAKA: true

InlayHints:
  Enabled: Yes
  BlockEnd: No
  Designators: No
  ParameterNames: Yes
  DeducedTypes: Yes
  TypeNameLimit: 0


CompileFlags:
  CompilationDatabase: <path to compilation database parent dir>
  Add:
    [
      -Wunused-function,
      -Wall,
      -Wextra,
      -Wshadow,
      -Wpedantic,
      -Werror=unused-variable,
      -Wexceptions,
      -Wno-switch-enum,
      -Wimplicit-fallthrough,
      -Wno-unknown-warning-option,
      --gcc-toolchain=<path to GGC>,
    ]
  Remove: -Werror

Diagnostics:
  UnusedIncludes: Strict
  ClangTidy:
    FastCheckFilter: Strict
    Add: [
        clang-diagnostic-*,
        clang-analyzer-*,
        modernize*,
        performance*,
        readability*,
        google-explicit-constructor,
        bugprone-exception-escape,
        bugprone-incorrect-enable-if,
        bugprone-inaccurate-erase,
        bugprone-redundant-branch-condition,
        bugprone-pointer-arithmetic-on-polymorphic-object,
        bugprone-parent-virtual-call,
        bugprone-unhandled-self-assignment,
        bugprone-unused-raii,
        bugprone-use-after-move,
        bugprone-unused-local-non-trivial-variable,
        cppcoreguidelines-slicing,
        misc-unused-parameters,
      ]
    Remove:
      [
        modernize-use-trailing-return-type,
        modernize-use-using,
        readability-avoid-unconditional-preprocessor-if,
        readability-identifier-length,
        readability-implicit-bool-conversion,
        readability-isolate-declaration,
        readability-magic-numbers,
        readability-qualified-auto,
        readability-suspicious-call-argument,

      ]
    CheckOptions:
      readability-function-cognitive-complexity.Threshold: 25,
      readability-function-cognitive-complexity.IgnoreMacros: true

Index:
    Background: Build
    StandardLibrary: No
```

I originally posted it as part of a question. Since I have not got any response, I am reposting it as enhancement request.

**Comment by HighCommander4** [comment]
I think passing `--compile-commands-dir=<dir>` as a command-line argument to clangd (with `<dir>` being the directory containing `compile_commands.json`) should result in everything being put in `<dir>/.cache/clangd` and nothing in `${HOME}/.cache/clangd`.

`CompilationDatabase: <dir>` in `.clangd` does not result in the same guarantee because `.clangd` files are scoped to only pertain to source files in the directory tree of the directory containing the `.clangd` file, and so e.g. system headers will fall outside of their scope.

The number of threads is also controlled with a command-line argument, `-j N`.

**Comment by soumen-ghosh** [comment]
I wanted to use `.clangd` file to make it IDE agnostic (ie it works same way for emacs or clion or vs code). Is there some other file that I can use instead of specifying it in something like `.vscode/settings.json` which will work only for vs code?

**Comment by HighCommander4** [comment]
> I wanted to use `.clangd` file to make it IDE agnostic (ie it works same way for emacs or clion or vs code). Is there some other file that I can use instead of specifying it in something like `.vscode/settings.json` which will work only for vs code?

Not that I can think of currently.

There is a proposal in https://github.com/clangd/clangd/discussions/907 about introducing a "strong workspace mode". One of its effects would be to make `CompilationDatabase: <dir>` behave entirely like `--compile-commands-dir=<dir>`, which seems like it would address this use case. (However, opting into strong workspace mode will likely still require client-specific configuration.)

> Also, what happens if my `compile_commands.json` in the sandbox is a soft-link to a file in the build area (different Unix path from my sandbox directory)?

I believe that works, that's fairly commonly done.

**Comment by soumen-ghosh** [comment]
> There is a proposal in #907 about introducing a "strong workspace mode". One of its effects would be to make `CompilationDatabase: <dir>` behave entirely like `--compile-commands-dir=<dir>`, which seems like it would address this use case. (However, opting into strong workspace mode will likely still require client-specific configuration.)

It looks like this #907 is more like what I also want. Maybe you can mark these as duplicate or link these requests. Regarding client-specific part of config vs generic part of config, if there is a way to have hierarchical config (similar to include for generic part in client config), it'll be helpful. I am not sure something like `Hierarchical configuration composable from multiple sources` from [hydra](https://hydra.cc/docs/intro/) will be of any help for this.

**Comment by soumen-ghosh** [comment]
> > Also, what happens if my `compile_commands.json` in the sandbox is a soft-link to a file in the build area (different Unix path from my sandbox directory)?
> 
> I believe that works, that's fairly commonly done.

Let me clarify it a bit more. We have a custom build system (different from plain make or cmake). And to avoid corrupting the sandbox, we create linked build area where all source files are symlinks to the ones from the sandbox and do the build there. As a result, the compilation database gets created in the linked build area. The file names in the compilation db are the soft-link each of which essentially points to corresponding file in the sandbox. Essentially, it looks like below -
```
<sandbox>
   |
   + compile_commands.json -> <build area>/compile_commands.json
```
Now going back to my question, if I create a soft-link in the sandbox pointing to the generated compilation db in the build area, will the indexing work? If no, what are other options that I can explore for this use model?

**Comment by HighCommander4** [comment]
> if I create a soft-link in the sandbox pointing to the generated compilation db in the build area, will the indexing work?

In this scenario, the index will be stored in `<sandbox>/.cache/clangd/index`.

**Comment by soumen-ghosh** [comment]
> In this scenario, the index will be stored in `<sandbox>/.cache/clangd/index`. Not sure if that's the behaviour you're looking for or not.
Yes. This is what I wanted as mentioned in the original question `I want indexes to be generated in the sandbox (project directory) `. This seems to be working for VS Code (as I have added `--compile-commands-dir=<dir>` in `.vscode/settings.json` through `"clangd.arguments"` key) but CLion or emacs (spacemacs with lsp) seems to be writing the index under `${HOME}/.cache/clangd/index`. Maybe, I need to find appropriate settings for those editors also. 

If .clangd is tied to workspace and takes highest precedence, one need not tweak things across multiple editors.
